pkg/e2e/operators/mnmo.go

```diff
@@ -0,0 +1,622 @@
+	// returning a non-nil error will quickly exit the test function failing the upstream test.
+	// returning a false, nil combination will retry the test
+	// returning a true, nil combination will exit the test successfully
+	test func([]*corev1.Node) (bool, error)
+}
+
+func waitFor(test MNMOTest, h *helper.H, machineSetName string) error {
+	var nodeListError error
+	for t := 0 * time.Second; t < test.timeout; t = t + test.interval {
+		nodeList, err := getNodesForMachineSet(h, machineSetName)
+		if nodeListError = err; err != nil {
+			// If there's some error getting the node list, persist it
+			// outside the loop in case it repeats.  We'll retry until the
+			// timeout, though. We reset nodeListError every time so in the
+			// case there's an error the first time but every other time it
+			// works we don't create a red herring.
+			time.Sleep(test.interval)
+			continue
+		}
+
+		if len(nodeList) < 1 {
+			// Make sure we're getting a list of nodes, otherwise
+			// just sleep and retry (otherwise we get a nil pointer panic)
+			time.Sleep(test.interval)
+			continue
+		}
+
+		pass, err := test.test(nodeList)
+		if err != nil {
+			// Test explicitly errored.  Exit out.
+			return err
+		}
+
+		if !pass {
+			// Test did not pass, retry after interval
+			time.Sleep(test.interval)
+			continue
+		}
+
+		// test has passed, exit out
+		return nil
+	}
+
+	// if listing the nodes returns an error consistently, let's return that,
+	// otherwise let's just say the test timed out, which means MNMO isn't working.
+	if nodeListError != nil {
+		return nodeListError
+	}
+	return fmt.Errorf("Test has timed out, something is wrong with MNMO or there is a network issue with OCM")
+}
+
+// Note these are ORDERED tests.  While ordering tests is generally a code smell, in this case
+// we don't want to do the expensive setup and teardown of the MachinePool, both in the cost of
+// $$ for the machinepool but also for the time it takes for it to come online.  So these tests
+// are Ordered to gain the most value, as they mimic a potential customer workflow over time
+
+// The only real drawback here to the ordered tests is that if the label tests fail, it won't
+// run the tests on the taint.
+var _ = ginkgo.Describe(mnmoOperatorTestName, ginkgo.Ordered, func() {
+	h := helper.New()
+	var (
+		ocm                *ocmprovider.OCMProvider
+		clusterClient      *ocmTypes.ClusterClient
+		clusterId          string
+		machinePoolBuilder *ocmTypes.MachinePoolBuilder
+		machineSetName     string
+	)
+
+	ginkgo.BeforeAll(func() {
+		var err error
+		ocm, err = ocmprovider.New()
+		Expect(err).NotTo(HaveOccurred())
+		Expect(ocm).NotTo(BeNil())
+
+		clusterId = viper.GetString(config.Cluster.ID)
+
+		// build the ocm client for this specific cluster
+		clusterClient = ocm.GetConnection().ClustersMgmt().V1().Clusters().Cluster(clusterId)
+
+		// build the cluster client to interact with things on-cluster with
+		h.Impersonate(rest.ImpersonationConfig{
```

Comment:
This part didn't work, when I tried to run them locally a few weeks ago, due to permission errors for the user.

This user was not allowed to list the machinesets, so I had to change it to this:

```
h.SetServiceAccount("system:serviceaccount:%s:cluster-admin")
```